### ISSUE TITLE
Reproducing the results in the paper

### ISSUE THREAD

**Issue by shadow2496** [rinongal/textual_inversion#92]
Hi, I'm trying to reproduce the paper results, especially the 'cat_statue' concept below in Figure 1, but it seems like extra information is needed to do that.
![image](https://user-images.githubusercontent.com/17617775/192135426-35cefa6b-22ec-4695-94fe-c40a629d459d.png)

### Inversion and Generation Settings
I'll describe my settings first.
- Input samples: following the images in the figure, I put 3 images ('2.jpg', '3.jpg', and '6.jpg') into the 'cat_statue_236' folder and used the folder as a training dataset. No preprocessing was applied.
![image](https://user-images.githubusercontent.com/17617775/192138415-1ed2c06c-d8ea-41f4-bb4f-913a8b880be2.png)
- Change in 'main.py': 'torch.backends.cudnn.deterministic = True' was added for reproducibility.
```
    ...
    cfgdir = os.path.join(logdir, "configs")
    seed_everything(opt.seed)
    torch.backends.cudnn.deterministic = True

    try:
    ...
```
- Inversion command: please check the command below. I set the initialization word as 'cat' and ran the command on 2 GPUs as written in the paper.
```
python main.py --base configs/latent-diffusion/txt2img-1p4B-finetune.yaml \
               -t \
               --actual_resume models/ldm/text2img-large/model.ckpt \
               -n init_cat_gpu2_s0023 \
               --gpus 0,1 \
               --data_root datasets/cat_statue_236 \
               --init_word cat \
               lightning.trainer.benchmark=False  # also for reproducibility
```
- Generation command: please check the command below. A seed argument was added for reproducibility, and all results were produced using 5,000 optimization steps following the implementation details in the paper. The attempted prompts were:
  - A photo of *
  - Banksy art of *
  - Painting of a * riding a dragon
  - Painting of two * fishing on a boat
  - A * themed lunchbox

```
python scripts/txt2img.py --ddim_eta 0.0 \
                          --n_samples 8 \
                          --n_iter 2 \
                          --scale 10.0 \
                          --ddim_steps 50 \
                          --embedding_path logs/cat_statue_236<now>_init_cat_gpu2_s0023/checkpoints/embeddings_gs-4999.pt \
                          --ckpt_path models/ldm/text2img-large/model.ckpt \
                          --prompt <prompt> \
                          --seed 97  # just a random number
```

### Generation Results
With the inversion and generation settings, I could obtain the results below. (up: my results, down: results from the supplementary)
- A photo of *
![image](https://user-images.githubusercontent.com/17617775/192269408-b3060a4b-9ece-416b-88e0-0536535d592f.png)
![image](https://user-images.githubusercontent.com/17617775/192269509-3ed84d59-aefa-430b-949f-6e9397f687fa.png)
- Banksy art of *
![image](https://user-images.githubusercontent.com/17617775/192270314-4ae75b6f-9578-484b-ac9e-1d2ae1f53600.png)
![image](https://user-images.githubusercontent.com/17617775/192269533-a471b99a-898a-4f46-b1be-1564388cf407.png)
- Painting of a * riding a dragon
![image](https://user-images.githubusercontent.com/17617775/192270422-f48aa9ba-265d-4201-a1ce-9da09e722fb2.png)
![image](https://user-images.githubusercontent.com/17617775/192269545-8dd9d9e6-de3c-4b1e-bc79-37b573121ac7.png)
- Painting of two * fishing on a boat
![image](https://user-images.githubusercontent.com/17617775/192270489-5049c6d3-1d51-4508-a887-54ca4203b465.png)
![image](https://user-images.githubusercontent.com/17617775/192269557-7eed372f-fe45-42b7-b772-672261c5a4f9.png)
- A * themed lunchbox
![image](https://user-images.githubusercontent.com/17617775/192270540-d5ea5efa-ae81-4d64-abe9-20128b12184d.png)
![image](https://user-images.githubusercontent.com/17617775/192269568-a5832ffa-a02c-46a6-801c-60b53dbb5eed.png)

### Questions
As you can see, my results are quite different from the ones in the paper. My questions are as follows.
1. Could you check if there is anything to change in the settings above?
2. The seed for the inversion is 23 by default. Should I change the seed number? It would be of great help if you could provide the seed for the paper results.
3. Other than the settings above and the inversion seed, are there any settings that I missed?

**Comment by shadow2496** [comment]
Oh, didn't know I should use all 7 images in the 'cat_statue' folder. I'll retrain as well.
And the batch size was 4 by default, which matches the paper.

Another question. For the inversion of the concepts other than 'cat_statue', did you also use all of the images in the corresponding google drive folder, not just the ones shown in the figure?

**Comment by rinongal** [comment]
Indeed. We note in some places in the papers that the figures show only exemplars from the full sets, but I'll make a note to better highlight it when we upload a revision. Regardless, you have the full sets in that drive. 

Some of the sets have associated copyrights so I didn't upload them, but if there's any particular set which is not up and you want the images for, let me know and I'll send it to you.

**Comment by rinongal** [comment]
Here we go:

Checkpoint:
[cat_26_09.zip](https://github.com/rinongal/textual_inversion/files/9647761/cat_5k_26_09.zip)

a painting of a * riding a dragon
![a-painting-of-a-%2A-riding-a-dragon](https://user-images.githubusercontent.com/19426368/192322423-0daf5765-98a8-412c-b970-4f1ffec20332.jpg)

a * themed lunchbox
![a-%2A-themed-lunchbox](https://user-images.githubusercontent.com/19426368/192322432-4e9a1cdb-a187-4232-91f3-5c87803b4026.jpg)

banksy art of *
![banksy-art-of-%2A](https://user-images.githubusercontent.com/19426368/192322437-33292dd9-27f7-4fbe-be01-d566caa29ad5.jpg)

a photo of *
![a-photo-of-%2A](https://user-images.githubusercontent.com/19426368/192322440-cc3708e9-96bf-4827-a055-58ab320e3d2f.jpg)

painting of two * fishing on a boat
![painting-of-two-%2A-fishing-on-a-boat](https://user-images.githubusercontent.com/19426368/192322443-efad3e3d-f507-4701-a4eb-ea250fb039dc.jpg)

I'm using the default seed (23). That said, I didn't clone a fresh repo and I have some local code changes (this path doesn't enter them, but they might be changing seeding results). Let me know if you still can't get it to work with the full set and I'll re-run on a fresh clone and also give you a seed so you can verify yourself.

**Comment by shadow2496** [comment]
Well, I retrained with the full set and these are the results I get.

- A photo of *
![image](https://user-images.githubusercontent.com/17617775/192520228-2756635c-b037-4d22-b3e1-15f62dc993fb.png)
- Banksy art of *
![image](https://user-images.githubusercontent.com/17617775/192520285-1ef9232f-e584-47ea-8de5-6c36f6b37a9e.png)
- Painting of a * riding a dragon
![image](https://user-images.githubusercontent.com/17617775/192520759-0ec2de5e-492e-4ca2-8ddc-7356563c8cd7.png)
- Painting of two * fishing on a boat
![image](https://user-images.githubusercontent.com/17617775/192520806-3bdeca69-a585-47ef-8f29-f9884a410da0.png)
- A * themed lunchbox
![image](https://user-images.githubusercontent.com/17617775/192520858-06e28acf-540a-402e-b4cd-c40fbcb96d43.png)

Still, my results need some improvements. Could you try the inversion on the fresh clone and see if the paper results are reproducible?

**Comment by rinongal** [comment]
Sure. Is the main.py change above the only adjustment you made? I'll do the same locally to make sure I'm getting your results with the default seed.

**Comment by rinongal** [comment]
Just a heads up that I haven't forgotten you. I'm also seeing an issue with a fresh install of everything + your modifications + new environment from scratch. 

I'm trying to track down the difference, but I'm doing a bunch of things in parallel so it might take a few days. 

If you need this urgently and want me to send you the closest image of the code to when we ran this experiment, let me know.  I ran that version on 2-3 seeds and it seems fine.

**Comment by ryanrussell** [comment]
@rinongal 

Thanks for sharing your checkpoint. Your example above made it crystal clear.

Just a note, using the latest from `rinongal/textual_inversion.git` on the `main` branch does not support the `seed` CLI arg.

```bash
txt2img.py: error: unrecognized arguments: --seed 97
```

Without that arg, it works great.

Is there a different branch for using `seed` via cli?

**Comment by rinongal** [comment]
No different branch using seed via CLI but it's trivial to add. You just copy the seed code from the stable_txt2img.py (adding it to the parser + the import / call to `seed_everything`). If you need help with that let me know and I'll update the repo.

**Comment by shadow2496** [comment]
@rinongal It would be great if I could use your code.
If you don't mind, could you send it to dadadak4@gmail.com?

**Comment by rinongal** [comment]
Sent. If anyone else needs the same code version before a fix is out, please let me know.

**Comment by shadow2496** [comment]
@rinongal Thank you very much.

**Comment by 1blackbar** [comment]
So the current code is unable to get the same editability? why is that ?

**Comment by rinongal** [comment]
@1blackbar 
Still trying to find where the issue stems from. At the moment this seems like an LDM only-issue, which likely came about when I added the partial SD support. I'm adding back my changes one-by-one to see where things break, but it's a lengthy process.

**Comment by shadow2496** [comment]
@rinongal I tried to find the cause on my end, and aside from the training templates, the training seems sensitive to the PIL version. Which PIL version did you use for the training?

**Comment by rinongal** [comment]
@shadow2496 Training templates do not appear to be the source. It was the first thing I verified when I pulled the old code.
 
I've got one environment running pillow 9.1.1, and one running 9.0.1. 
I'm not seeing meaningfully different results in one compared to the other.

**Comment by shadow2496** [comment]
Hmm, I couldn't find other differences that affect the training than the conda environment and the training templates from the code you sent.
The problem is I can't reproduce the results even from your code. Maybe the issue comes from using different GPUs? (I'm using TITAN RTX)

- A photo of *. (training templates included periods)
![image](https://user-images.githubusercontent.com/17617775/194395404-8a550476-5115-4fdf-b341-af571d0d1236.png)
- Banksy art of *. (training templates included periods)
![image](https://user-images.githubusercontent.com/17617775/194395506-33de3098-be0f-4753-8b76-bd29ea53ef95.png)
- Painting of a * riding a dragon
![image](https://user-images.githubusercontent.com/17617775/194396905-dbd25086-d817-494c-ae3e-cdb2ce3f89b6.png)
- Painting of two * fishing on a boat
![image](https://user-images.githubusercontent.com/17617775/194397124-911c905e-acf0-49be-90a9-216619ec62b3.png)
- A * themed lunchbox
![image](https://user-images.githubusercontent.com/17617775/194397253-990b03d2-a49b-455e-96bd-bcd22421f2d7.png)

**Comment by shadow2496** [comment]
Just to make sure the difference in conda environments is not an issue, could you send the yaml file from the `conda env export > environments.yaml` command?

**Comment by rinongal** [comment]
[environments_freeze.txt](https://github.com/rinongal/textual_inversion/files/9747878/environments_freeze.txt)
Here you go. I changed the extension to .txt. since github won't let me upload a yaml. Just swap it back.

And here's the settings from the original run that produced the paper results. It's using 4 workers, but the checkpoint I sent here was made using 2 so that shouldn't be an issue.

[cat2022-07-03T12-29-05-project.txt](https://github.com/rinongal/textual_inversion/files/9747946/cat2022-07-03T12-29-05-project.txt)

I'm honestly still not sure what's causing the discrepancy here. I'll try setting things up in a colab so we can make sure we're getting synchronized results and work from there.

**Comment by CrossLee1** [comment]
@shadow2496 do you succeed to reproduce the result.? I met the same problem with you ..

**Comment by CrossLee1** [comment]
@rinongal 
one more question:
in the enviroments_freeze.txt you provide above,  transformers==4.3.1
but in https://github.com/rinongal/textual_inversion/blob/main/environment.yaml, transformers==4.18.0
which one is correct

**Comment by rinongal** [comment]
@CrossLee1 The enviroments_freeze.txt above is the one I used locally to train the cat checkpoint in the beginning of this thread. The one in the repo is just updated to reflect Stable-Diffusion requirements. Unfortunately, we made quite a few changes to add support for SD and since training here is seed sensitive, these tend to impact the results as well.

I'm trying to find the time to look deeper into this or set up a colab where I can search for a seed that gives close results in a shared, consistent environment. Unfortunately I'm a bit bogged down with other things so until I get around to that, 'seed sensitivity' is the best answer I can give you.

**Comment by rinongal** [comment]
@CrossLee1 The enviroments_freeze.txt above is the one I used locally to train the cat checkpoint in the beginning of this thread. The one in the repo is just updated to reflect Stable-Diffusion requirements. Unfortunately, we made quite a few changes to add support for SD and since training here is seed sensitive, these tend to impact the results as well.

I'm trying to find the time to look deeper into this or set up a colab where I can search for a seed that gives close results in a shared, consistent environment. Unfortunately I'm a bit bogged down with other things so until I get around to that, 'seed sensitivity' is the best answer I can give you.

**Comment by Delicious-Bitter-Melon** [comment]
@shadow2496 @CrossLee1  do you succeed to reproduce the result? I met the same problem with you .
